pilot/pkg/proxy/envoy/v1/authentication.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// AuthN filter configuration
+
+package v1
+
+import (
+	"fmt"
+
+	authn "istio.io/api/authentication/v1alpha1"
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// AuthenticationFilterName is the name for filter
+	AuthenticationFilterName = "isiot_authn"
+)
+
+// getConsolidateAuthenticationPolicy returns the authentication policy for
+// service specified by hostname and port, if defined.
+// If not, it generates and output a policy that is equivalent to the legacy flag
+// and/or service annotation. Once these legacy flags/config deprecated,
+// this function can be placed by a call to store.AuthenticationPolicyByDestination
+// directly.
+func getConsolidateAuthenticationPolicy(mesh *meshconfig.MeshConfig, store model.IstioConfigStore, hostname string, port *model.Port) *authn.Policy {
+	config := store.AuthenticationPolicyByDestination(hostname, port)
+	if config == nil {
+		legacyPolicy := consolidateAuthPolicy(mesh, port.AuthenticationPolicy)
+		log.Debugf("No authentication policy found for  %s:%d. Fallback to legacy authentication mode %v\n",
+			hostname, port.Port, legacyPolicy)
+		return legacyAuthenticationPolicyToPolicy(legacyPolicy)
+	}
+
+	return config.Spec.(*authn.Policy)
+}
+
+// consolidateAuthPolicy returns service auth policy, if it's not INHERIT. Else,
+// returns mesh policy.
+func consolidateAuthPolicy(mesh *meshconfig.MeshConfig,
+	serviceAuthPolicy meshconfig.AuthenticationPolicy) meshconfig.AuthenticationPolicy {
+	if serviceAuthPolicy != meshconfig.AuthenticationPolicy_INHERIT {
+		return serviceAuthPolicy
+	}
+	// TODO: use AuthenticationPolicy for mesh policy and remove this conversion
+	switch mesh.AuthPolicy {
+	case meshconfig.MeshConfig_MUTUAL_TLS:
+		return meshconfig.AuthenticationPolicy_MUTUAL_TLS
+	case meshconfig.MeshConfig_NONE:
+		return meshconfig.AuthenticationPolicy_NONE
+	default:
+		// Never get here, there are no other enum value for mesh.AuthPolicy.
+		panic(fmt.Sprintf("Unknown mesh auth policy: %v\n", mesh.AuthPolicy))
```

Comment:
Really? This is bad. Why not just error out?

Reply:
This is moved from config.go. 
Basically in the mesh config, we have several enum for authn policy (sigh), and some restriction that a certain enum cannot be used for mesh (aka global setting). I know it's ugly, and hope we can clean them up with policy.